Linked Keywords

proteusvacuum opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->
https://trello.com/c/DT4jCByP/117-add-sms-keywords-to-linked-project-spaces

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This doesn't handle remote links.
It prevents linking keywords that notify groups, as we don't maintain a mapping for those between projects.

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->
Linked domains

##### RISK ASSESSMENT / QA PLAN
<!-- Does this need QA before or after merge? Link QA ticket. -->
QA is needed

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

proteusvacuum committed:
Copy keyword into linked domain

proteusvacuum committed:
Add modal to select keywords to copy

- There are only a few hundred keywords on HQ (for all projects), so fetching
all keywords in a domain should not be a performance hit

proteusvacuum committed:
Update linked keywords

proteusvacuum committed:
Add keyword name into views

proteusvacuum committed:
Add keywords to release manager

proteusvacuum committed:
Only show link keyword button if it will do anything

proteusvacuum committed:
Only link linkable keywords and show error messages

proteusvacuum committed:
Raise error for remote links

proteusvacuum committed:
More error handling

proteusvacuum committed:
Lint

orangejenny commented:
I haven't read through all of this yet, but it looks like it includes an explicit linking step. I was imagining that keywords would just all be pulled down, similarly to user roles. Does that not work out well?

proteusvacuum commented:
@orangejenny Is that the desired behaviour? I could change it to do that, but I assumed that people might want to be more selective (e.g. if they had test keywords, or keywords that were linked to apps that were not pushed, etc.).

orangejenny commented:
@proteusvacuum Hmm, I see your point. I'm not sure, I think it's a question for product.

orangejenny committed:
Merge branch 'master' into fr/linked-domain-keywords

orangejenny committed:
Incremented migration

orangejenny committed:
Fixed name of previous migration

orangejenny commented on corehq/apps/linked_domain/keywords.py:
```diff
@@ -10,24 +13,52 @@ def create_linked_keyword(domain_link, keyword_id):
+
+
+@quickcache(vary_on=['domain'], skip_arg=lambda _: settings.UNIT_TESTING, timeout=5 * 60)
+def get_master_app_to_linked_app(domain):
```
This will be subject to the same issue as reports, where it doesn't support linked apps with multiple downstream apps, right? What if this function just discarded any apps that share the same family_id, would that make this fail more quickly/loudly? Or threw an exception to alert the user that this isn't a supported workflow?

orangejenny replied:
Made a note on https://dimagi-dev.atlassian.net/browse/USH-167, doesn't need to block this PR.

orangejenny commented on corehq/apps/reminders/views.py:
```diff
@@ -330,3 +336,37 @@ def get_deleted_item_data(self, item_id):
+            request,
+            _(f"Successfully linked and copied to {', '.join(successes)}. "))
+    if failures:
+        messages.error(request, _(f"Due to errors, the report was not copied to {', '.join(failures)}"))
```
"report" => "keyword"

orangejenny replied:
fixed in later commit

orangejenny commented on corehq/apps/linked_domain/tasks.py:
```diff
@@ -166,6 +169,22 @@ def _release_case_search(self, domain_link, model, user):
+    def _release_keyword(self, domain_link, model):
+        master_id = model['detail']['keyword_id']
+        try:
+            linked_keyword_id = Keyword.objects.values_list('id', flat=True).get(domain=domain_link.linked_domain, master_id=master_id)
```
What's the purpose of the `values_list` call?

orangejenny replied:
fixed in later commit

orangejenny commented on corehq/apps/reminders/templates/reminders/keyword_list.html:
```diff
@@ -21,7 +21,11 @@ <h2>{% trans 'Manage Keywords' %}</h2>
       {% trans 'Add Structured Keyword' %}
     </a>
-    {% include "reminders/link_keywords_modal.html" %}
+    {% if request|toggle_enabled:"LINKED_DOMAINS" %}
```
This toggle is on for both upstream and downstream domains, but the button should only show for upstream domains, right?

orangejenny replied:
Nevermind, I must have missed the `if linked_domains` check below.

orangejenny reviewed (commented):
Just minor feedback.

orangejenny commented on corehq/apps/reminders/templates/reminders/link_keywords_modal.html:
```diff
@@ -0,0 +1,56 @@
+            <div class="controls col-xs-10">
+              <select name="to_domains" class="form-control" required="" id="keyword-selector" multiple>
+                {% for keyword in all_keywords %}
+                  <option value="{{ keyword.couch_id }}">
```
Ignore, you got rid of this later.

proteusvacuum committed:
Rename model detail variable for clarity

proteusvacuum committed:
Show domains once

orangejenny committed:
Merge branch 'master' into fr/linked-domain-keywords

orangejenny committed:
Added merge migration

orangejenny committed:
Updated unlinked app error to be more specific

orangejenny committed:
Updated to reset all keyword actions on each update

Keyword actions can't be linked by id, because the upstream keyword action ids aren't stable.

orangejenny committed:
Removed new keyword if creating keyword fails to create actions

orangejenny committed:
Renamed master_id to upstream_id

orangejenny committed:
Fixed create_linked_keyword to re-raise any error in updating actions

orangejenny committed:
Updated exceptions with error text passed to user

orangejenny commented on corehq/apps/linked_domain/keywords.py:
```diff
@@ -10,24 +13,52 @@ def create_linked_keyword(domain_link, keyword_id):
 
     keyword_actions = keyword.keywordaction_set.all()
 
+    keyword.master_id = keyword.id
```
Changed in https://github.com/dimagi/commcare-hq/pull/28432/commits/c06f33c1f9d4025385530619ac279bd9bfc99a6a

orangejenny committed:
Added initiator_doc_type_filter to updated properties

orangejenny commented on corehq/apps/linked_domain/keywords.py:
```diff
@@ -10,24 +13,52 @@ def create_linked_keyword(domain_link, keyword_id):
+    linked_keyword = Keyword.objects.get(id=keyword_id)
+    master_keyword = Keyword.objects.get(id=linked_keyword.master_id)
+
+    for prop in ['keyword', 'description', 'delimiter', 'override_open_sessions']:
```
Added in abecb5b564c

orangejenny committed:
Merge branch 'master' into fr/linked-domain-keywords
